File: CS/DevExpress.OutlookInspiredApp.Wpf/ViewModels/EmployeeTask/EmployeeTaskDetailsCollectionViewModel.cs
using System;
using System.Linq;
using DevExpress.Data.Filtering;
using DevExpress.Data.Utils;
using DevExpress.DevAV.Common.DataModel;
using DevExpress.DevAV.Common.ViewModel;
using DevExpress.DevAV.DevAVDbDataModel;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;

namespace DevExpress.DevAV.ViewModels {
    public class EmployeeTaskDetailsCollectionViewModel : CollectionViewModel<EmployeeTask, long, IDevAVDbUnitOfWork> {
        public static EmployeeTaskDetailsCollectionViewModel Create(IUnitOfWorkFactory<IDevAVDbUnitOfWork> unitOfWorkFactory = null) {
            return ViewModelSource.Create(() => new EmployeeTaskDetailsCollectionViewModel(unitOfWorkFactory));
        }
        protected EmployeeTaskDetailsCollectionViewModel(IUnitOfWorkFactory<IDevAVDbUnitOfWork> unitOfWorkFactory = null)
            : base(unitOfWorkFactory ?? UnitOfWorkSource.GetUnitOfWorkFactory(), x => x.Tasks, query => query.OrderByDescending(x => x.DueDate)) {
        }
        public IDocumentManagerService AssignTaskService { get { return this.GetService<IDocumentManagerService>("EmployeeAssignService"); } }

        public override void Edit(EmployeeTask projectionEntity) {
            base.Edit(projectionEntity);
            UpdateFilter();
        }
        public override void Delete(EmployeeTask projectionEntity) {
            base.Delete(projectionEntity);
            UpdateFilter();
        }
        public override void New() {
            GetDocumentManagerService().ShowNewEntityDocument<EmployeeTask>(this);
            UpdateFilter();
        }
        public virtual void AssignTask(EmployeeTask task) {
            IDocument document = AssignTaskService.CreateDocument("EmployeeAssignView", task.Id, this);
            if(document != null)
                document.Show();
            UpdateFilter();
        }
        public virtual bool CanAssignTask(EmployeeTask task) {
            return !IsLoading && task != null;
        }
        public void UpdateFilter() {
            var viewModel = this.GetParentViewModel<EmployeeViewModel>();
            var unitOfWork = unitOfWorkFactory.CreateUnitOfWork();
            var actualEntity = unitOfWork.Employees.FirstOrDefault(x => x.Id == viewModel.Entity.Id);
            if(viewModel != null)
                FilterExpression = CriteriaOperatorToExpressionConverter.GetGenericWhere<EmployeeTask>(CriteriaOperator.Parse(EmployeeTaskCollectionViewModel
                    .TasksToFilterCriteriaConverter((actualEntity == null) ? null : actualEntity.AssignedEmployeeTasks)));
        }
    }
}

File: VB/DevExpress.OutlookInspiredApp.Wpf/ViewModels/EmployeeTask/EmployeeTaskDetailsCollectionViewModel.vb
Imports System
Imports System.Linq
Imports DevExpress.Data.Filtering
Imports DevExpress.Data.Utils
Imports DevExpress.DevAV.Common.DataModel
Imports DevExpress.DevAV.Common.ViewModel
Imports DevExpress.DevAV.DevAVDbDataModel
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO

Namespace DevExpress.DevAV.ViewModels
    Public Class EmployeeTaskDetailsCollectionViewModel
        Inherits CollectionViewModel(Of EmployeeTask, Long, IDevAVDbUnitOfWork)

        Public Shared Function Create(Optional ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of IDevAVDbUnitOfWork) = Nothing) As EmployeeTaskDetailsCollectionViewModel
            Return ViewModelSource.Create(Function() New EmployeeTaskDetailsCollectionViewModel(unitOfWorkFactory))
        End Function
        Protected Sub New(Optional ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of IDevAVDbUnitOfWork) = Nothing)
            MyBase.New(If(unitOfWorkFactory, UnitOfWorkSource.GetUnitOfWorkFactory()), Function(x) x.Tasks, Function(query) query.OrderByDescending(Function(x) x.DueDate))
        End Sub
        Public ReadOnly Property AssignTaskService() As IDocumentManagerService
            Get
                Return Me.GetService(Of IDocumentManagerService)("EmployeeAssignService")
            End Get
        End Property

        Public Overrides Sub Edit(ByVal projectionEntity As EmployeeTask)
            MyBase.Edit(projectionEntity)
            UpdateFilter()
        End Sub
        Public Overrides Sub Delete(ByVal projectionEntity As EmployeeTask)
            MyBase.Delete(projectionEntity)
            UpdateFilter()
        End Sub
        Public Overrides Sub [New]()
            GetDocumentManagerService().ShowNewEntityDocument(Of EmployeeTask)(Me)
            UpdateFilter()
        End Sub
        Public Overridable Sub AssignTask(ByVal task As EmployeeTask)
            Dim document As IDocument = AssignTaskService.CreateDocument("EmployeeAssignView", task.Id, Me)
            If document IsNot Nothing Then
                document.Show()
            End If
            UpdateFilter()
        End Sub
        Public Overridable Function CanAssignTask(ByVal task As EmployeeTask) As Boolean
            Return (Not IsLoading) AndAlso task IsNot Nothing
        End Function
        Public Sub UpdateFilter()
            Dim viewModel = Me.GetParentViewModel(Of EmployeeViewModel)()
            Dim unitOfWork = unitOfWorkFactory.CreateUnitOfWork()
            Dim actualEntity = unitOfWork.Employees.FirstOrDefault(Function(x) x.Id = viewModel.Entity.Id)
            If viewModel IsNot Nothing Then
                FilterExpression = CriteriaOperatorToExpressionConverter.GetGenericWhere(Of EmployeeTask)(CriteriaOperator.Parse(EmployeeTaskCollectionViewModel.TasksToFilterCriteriaConverter(If(actualEntity Is Nothing, Nothing, actualEntity.AssignedEmployeeTasks))))
            End If
        End Sub
    End Class
End Namespace
